src/main/scala/org/ergoplatform/nodeView/history/storage/modifierprocessors/HeadersProcessor.scala

```diff
@@ -267,18 +267,25 @@ trait HeadersProcessor extends ToDownloadProcessor with ScorexLogging with Score
       // Set difficulty for version 2 activation height (where specific difficulty is needed due to PoW change)
       settings.chainSettings.initialDifficultyVersion2
     } else {
-      //todo: it is slow to read thousands headers from database for each header
-      //todo; consider caching here
-      //todo: https://github.com/ergoplatform/ergo/issues/872
       val parentHeight = parent.height
-      val heights = difficultyCalculator.previousHeadersRequiredForRecalculation(parentHeight + 1)
-        .ensuring(_.last == parentHeight)
-      if (heights.lengthCompare(1) == 0) {
-        difficultyCalculator.calculate(Seq(parent))
+
+      // Difficulty readjustment happens on the first block of an epoch (height % epochLength == 1)
+      if(parentHeight % settings.chainSettings.epochLength == 0) {
+        //todo: it is slow to read thousands headers from database for each header
+        //todo; consider caching here
+        //todo: https://github.com/ergoplatform/ergo/issues/872
+        val heights = difficultyCalculator.previousHeadersRequiredForRecalculation(parentHeight + 1)
+          .ensuring(_.last == parentHeight)
+        if (heights.lengthCompare(1) == 0) {
+          difficultyCalculator.calculate(Seq(parent))
```

Comment:
Is Seq->Array replacement appreciate  here?

Reply:
Fixed, in #1618